source/functions/test.txt

```diff
@@ -0,0 +1,313 @@
+
+.. code-block:: js
+
+   global.context = {
+     // whichever global context methods you want to mock.
+     // 'services', 'functions', values, etc.
+   }
+
+   // you can also mock other Functions global modules
+   global.BSON = {
+     // mock methods
+   }
+
+You may also want to declare and remove these mocks in setup and teardown blocks
+so that they do not pollute the global namespace.
+
+.. code-block:: js
+   :emphasize-lines: 2-6, 9-11
+
+    // adds context mock to global namespace before each test
+    beforeEach(() => {
+      global.context = {
+        // your mocking services
+      };
+    });
+
+    // removes context from global namespace after each test
+    afterEach(() => {
+      delete global.context;
+    });
+
+    test("should perform operation using App Services globals", () => {
+      // test function that uses context
+    });
+
+.. example:: Mocking a Function that accesses ``context``
+
+   The function in this example accesses an App Services Value and returns it.
+
+   .. literalinclude:: /examples/generated/functions/helloWithValue.snippet.hello-with-context-value.js
+      :language: js
+      :caption: helloWithValue.js
+      :emphasize-lines: 2
+
+   Now create a test file ``helloWithValue.test.js``.
+   The test file contains the following:
+
+   - Import the function exported from ``helloWithValue.js``.
+   - A mock of ``context.values.get()``. Wrap the mock in set up and tear down blocks
+     so that it does not pollute the global namespace.
+   - A test of the imported function that uses the mock.
+
+   .. literalinclude:: /examples/generated/functions/helloWithValue.test.snippet.test-with-mocking.js
+      :language: js
+      :caption: helloWithValue.test.js
+
+.. _integration-test-functions:
+
+Integration Tests for Functions
+-------------------------------
+
+You should perform integration tests on all Functions before deploying
+them to production environments. This is especially important because
+the Atlas Function JavaScript runtime differs from the standard Node.js runtime.
+Unexpected errors can occur if you do not test functions deployed to App Services.
+
+There is no single way to write integration tests for Functions.
+As Functions can be used in a variety of different contexts for different purposes,
+each use case requires a different integration test strategy.
+
+For example, the way you create an integration test for a Function that you
+invoke from a Realm client SDK is different from the way you test
+a GraphQL Custom Resolver Function, which is different from how you would
+test a Database Trigger Function.
+
+However, there are some general steps that you can take to writing integration tests
+for Functions. On a high level these steps are:
+
+#. Create a testing App Services App with the same configuration as your production
+   App Services App.
+#. Write integration tests that use the Realm client SDKs to interact with the test App.
```

Comment:
Interesting definition of integration test here as a full end-to-end client-to-backend interaction. I think I would personally define "Function integration tests" as "tests where the *Function* actually interacts with a live testing backend"... I hadn't even thought about the client side. Personally I'd say that the client bit belongs in client SDK testing guides, since that's mostly about passing values correctly to the backend, as opposed to significant logic. But since the App is a full backend, perhaps there's more logic involved there than I was thinking.

I would recommend covering some of the "Function communicates with a live testing environment" bits here as well, though -- maybe you could demonstrate communicating with an external service like twilio in an integration test. Or maybe that's a future ticket. Up to you.

Reply:
For clarity, I look at the spectrum of integration testing for an App like this:

```
client SDKs -> App -> other backend stuff
```

I think this explanation does a good job of explaining how you'd integration test `client SDKs -> App`, but I think we could add a little more info about how you might integration test `App -> other backend stuff`.

Reply:
> Function communicates with a live testing environment

yes i see your point. in the below example i do use a realm sdk (as it provides an easy way to interact w atlas). but you're correct that the core idea is interacting w in a live env. 

> could add a little more info about how you might integration test App -> other backend stuff.

this is a tricky one b/c there's really no single way. this would require a different guide for each function use case scenario..and maybe those guides should exist as well. but those would be in different sections, ie page(s) for trigger types, page for graphql custom resolvers, page for data api custom endpoints, etc.